src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,272 @@
+        assignedLabels.addAll(initialLabels);
+        assignedLabels.addAll(addedLabels);
+        assignedLabels.removeAll(removedLabels);
+        return assignedLabels;
+    }
+
+    public List<String> getInitialLabels() {
+        return convertToList(initialLabels);
+    }
+
+    public List<String> getRemovedLabels() {
+        return convertToList(removedLabels);
+    }
+
+    public List<String> getAddedLabels() {
+        return convertToList(addedLabels);
+    }
+
+    public Optional<String> getCurrentSuggestion() {
+        if (currentSuggestionIndex.isPresent() && isValidSuggestionIndex()) {
+            return Optional.of(getSuggestedLabel());
+        }
+        return Optional.empty();
+    }
+
+    public List<String> getMatchedLabels() {
+        return matchedLabels;
+    }
+
+    /*
+     * Helper functions
+     */
+
+    private String getMatchedLabelName(Set<String> repoLabels, String keyword) {
+        List<String> newMatchedLabels = new ArrayList<>();
+        newMatchedLabels.addAll(repoLabels);
+        newMatchedLabels = filterByName(newMatchedLabels, getName(keyword));
+        newMatchedLabels = filterByGroup(newMatchedLabels, getGroup(keyword));
+        return newMatchedLabels.get(0);
+    }
+
+    private boolean hasExactlyOneMatchedLabel(Set<String> repoLabels, String keyword) {
+        List<String> newMatchedLabels = new ArrayList<>();
+        newMatchedLabels.addAll(repoLabels);
+        newMatchedLabels = filterByName(newMatchedLabels, getName(keyword));
+        newMatchedLabels = filterByGroup(newMatchedLabels, getGroup(keyword));
+        return newMatchedLabels.size() == 1;
+    }
+
+    private void removeConflictingLabels(String name) {
+        if (hasExclusiveGroup(name)) {
+            String group = getGroup(name);
+            // Remove from addedLabels
+            addedLabels = addedLabels.stream()
+                    .filter(label -> {
+                        String labelGroup = getGroup(label);
+                        return !labelGroup.equals(group);
+                    })
+                    .collect(Collectors.toSet());
+
+            // Add to removedLabels all initialLabels that have conflicting group
+            removedLabels.addAll(initialLabels.stream()
+                    .filter(label -> {
+                        String labelGroup = getGroup(label);
+                        return labelGroup.equals(group) && !removedLabels.contains(name);
+                    })
+                    .collect(Collectors.toSet()));
+        }
+    }
+
+    private boolean hasExclusiveGroup(String name) {
+        return TurboLabel.getDelimiter(name).isPresent() && TurboLabel.getDelimiter(name).get().equals(".");
+    }
+
+    private List<String> filterByName(List<String> repoLabels, String labelName) {
+        return repoLabels
+                .stream()
+                .filter(name -> Utility.containsIgnoreCase(getName(name), labelName))
```

Comment:
Maybe these methods could take and return `Stream<String>`. We don't need to build up the intermediate data structures, and they could be called by the new method which matches labels.